### PR TITLE
Encryption

### PR DESCRIPTION
**Pull request by adamsclafani**
Initial start with adding encryption. 
As of now, the "client" and the "server" are deriving the key each time using pbkdf2 at 100000 iterations each time which would probably slow down the benchmarks a bit more

### CODE REVIEW

**Comment by nzola** [comment]
Hi Adam,

What does this means in simple terms?

Thanks,
Nzola

**Comment by adamsclafani** [comment]
Hey!

At the moment, on each test, a key is being derived from a password, verified, then re-derived on the "client" side. This means, for each test, the key-derivation algorithm that generates a key to encrypt/decrypt the data, is being ran each time. To my understanding, the key needs to be generated, verified and derived only once in order to better mimic the real world usage.

**Comment by adamsclafani** [comment]
Either way though, I'm not actually too sure if this would change things that much, and from what I've seen, I don't think it would really make that big of a difference in terms of speeding things up.

**Comment by nzola** [comment]
Thanks Adam!

- Nzola

**Comment by adamsclafani** [comment]
As far as the extra size overhead goes, as of now, it should only be an extra 80 bytes per file

File: benchmarks/__main__.py [hunk 1]
```diff
@@ -81,17 +86,20 @@ def run_benchmarks(emails, results_dir, incremental):
         try:
             with Timer.timeit() as read_timer:
                 with compressor.open_read(outpath) as fobj:
-                    for _ in serializer.deserialize(fobj):
-                        pass
+                    with encryptor.deserialize(fobj) as dec:
```
Comment (c-w) [inline benchmarks/__main__.py]: From my understanding, modern encryption transforms the incoming bytes into an essentially random byte string (e.g. to prevent statistical frequency analysis attacks). As such, I believe the encryption should happen after the compression since otherwise the compression won't have enough signal/patterns in the encrypted byte string to produce good results. Am I mistaken here?

File: benchmarks/encryption.py [hunk 2]
```diff
@@ -0,0 +1,330 @@
+from abc import ABC
+from typing import IO
+from typing import Iterable
+from os import urandom
+from contextlib import contextmanager
+from tempfile import NamedTemporaryFile
+from base64 import b64encode
+
+
```
Comment (c-w) [inline benchmarks/encryption.py:9]: Nitpick: only one newline between import blocks. I'll add an import style linter for the project so that we're consistent with webapp and cloudserver.

File: benchmarks/encryption.py [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+        iv_sent = False
+
+        for raw in data:
+            out = encryptor.update(raw)
+
+            if not iv_sent:
+                out = self._iv + self._salt + self._hmac_salt + out
+                iv_sent = True
+
+            h.update(out)
```
Comment (c-w) [inline benchmarks/encryption.py:109]: Could you expand on this variable name to make the code more explicit?

File: benchmarks/encryption.py [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+            self._derive,
+            self._backend,
+            salt
+        )
+
+        return key_derive()
+
+    def _return_chunks(self, data, minsize=48):
+
+        cur, ahead = tee(data)
```
Comment (c-w) [inline benchmarks/encryption.py]: Is there any way to avoid using tee? From my understand, tee may require quite a bit of overhead. From the [stdlib docs on tee](https://docs.python.org/3/library/itertools.html#itertools.tee):

> This itertool may require significant auxiliary storage (depending on how much temporary data needs to be stored). In general, if one iterator uses most or all of the data before another iterator starts, it is faster to use list() instead of tee().

File: benchmarks/encryption.py [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+        ahead_val = next(ahead, None)
+
+        while ahead_val is not None:
+
+            bytes = next(cur, None)
```
Comment (c-w) [inline benchmarks/encryption.py]: `bytes` is a builtin function so using it as a name should be avoided since it'd shadow the builtin.

File: benchmarks/encryption.py [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+            default_backend
+        )
+
+        try:
+            with NamedTemporaryFile() as temp:
+                for item in decrypt(fobj):
+                    temp.write(item)
+                yield temp
+
+        finally:
```
Comment (c-w) [inline benchmarks/encryption.py]: Is this finally block (and the one above) required since it doesn't execute any code?

**Review by c-w — CHANGES_REQUESTED** [review]
Great start. Thanks so much for implementing this. The low size overhead is amazing.

A couple of comments + please also take a look at the flake8 issues:

flake8 lint failures

File: benchmarks/encryption.py (hunk 2 above)
Comment (c-w) [reply benchmarks/encryption.py:9]: The linter is now added, so if you rebase or merge your branch onto master then Travis will run isort as part of the CI.

File: benchmarks/__main__.py (hunk 1 above)
Comment (adamsclafani) [reply benchmarks/__main__.py]: Hmmm I actually have no Idea but that makes sense! I'll look more into this!

File: benchmarks/encryption.py (hunk 2 above)
Comment (adamsclafani) [reply benchmarks/encryption.py:9]: Ok awesome! I think it's about time for me to add an auto pep8 linter to my text editor as well 😅

File: benchmarks/encryption.py (hunk 3 above)
Comment (adamsclafani) [reply benchmarks/encryption.py:109]: For sure, will do!

File: benchmarks/encryption.py (hunk 4 above)
Comment (adamsclafani) [reply benchmarks/encryption.py]: Ah I see. Interesting. I'll see what I can do!

File: benchmarks/encryption.py (hunk 5 above)
Comment (adamsclafani) [reply benchmarks/encryption.py]: My bad. Will change.

File: benchmarks/encryption.py (hunk 6 above)
Comment (adamsclafani) [reply benchmarks/encryption.py]: Not needed at all. Will remove!

File: benchmarks/__main__.py (hunk 1 above)
Comment (adamsclafani) [reply benchmarks/__main__.py]: Just looked into this and yes, compression before encryption is the recommended course of action.

**Comment by adamsclafani** [comment]
Made a few minor changes. Still have to work out how I want to implement encryption after compression.

**Comment by adamsclafani** [comment]
Encryption after compression added. Some new functions had to be added to the compression bit in order to support passing a file object instead of a path.

File: .travis.yml [hunk 7]
```diff
@@ -7,10 +7,11 @@ python:
   - "3.6"
 
 before_script:
-  - pip install flake8
+  - pip install flake8 isort
```
Comment (c-w) [inline .travis.yml]: Please rebase to master to get rid of these spurious additions.

File: benchmarks/__main__.py [hunk 8]
```diff
@@ -47,31 +49,35 @@ def load_samples(zip_url, inputs_dir, exclude_attachments):
...
 
         if incremental and isfile(outpath):
             continue
 
         try:
             with Timer.timeit() as write_timer:
-                with compressor.open_write(outpath) as fobj:
-                    serializer.serialize(iter(emails), fobj)
+                with open(outpath, "wb") as fobj:
+                    with encryptor.encrypt(fobj) as enc:
```
Comment (c-w) [inline benchmarks/__main__.py]: I'm having a hard time figuring out how the encryptor works so I tried adding a unit test like this:

```patch
diff --git a/benchmarks/tests.py b/benchmarks/tests.py
index 653df57..4bb6ec7 100644
--- a/benchmarks/tests.py
+++ b/benchmarks/tests.py
@@ -6,15 +6,31 @@ from tempfile import mkstemp
 from unittest import TestCase
 
 from benchmarks.compression import get_all as compressors
+from benchmarks.encryption import get_all as encryptors
 from benchmarks.serialization import get_all as serializers
 
 
-class CompressionTests(TestCase):
+class TempfilesTestCase(TestCase):
+    def setUp(self):
+        self.temp_paths = []
+
+    def tearDown(self):
+        for temp_path in self.temp_paths:
+            remove(temp_path)
+
+    def given_tempfile(self, extension):
+        fd, temp_path = mkstemp(suffix=extension)
+        close(fd)
+        self.temp_paths.append(temp_path)
+        return temp_path
+
+
+class CompressionTests(TempfilesTestCase):
     def test_roundtrip(self):
         for compressor in compressors():
             with self.subTest(compressor=compressor):
                 expected = b'test content'
-                path = self.given_tempfile(compressor)
+                path = self.given_tempfile(compressor.extension)
 
                 with compressor.open_write(path) as compressed:
                     compressed.write(expected)
@@ -24,19 +40,6 @@ class CompressionTests(TestCase):
 
                 self.assertEqual(actual, expected)
 
-    def setUp(self):
-        self.temp_paths = []
-
-    def tearDown(self):
-        for temp_path in self.temp_paths:
-            remove(temp_path)
-
-    def given_tempfile(self, compressor):
-        fd, temp_path = mkstemp(suffix=compressor.extension)
-        close(fd)
-        self.temp_paths.append(temp_path)
-        return temp_path
-
 
 class SerializationTests(TestCase):
     def test_roundtrip(self):
@@ -57,6 +60,26 @@ class SerializationTests(TestCase):
                 self.assertListEqual(actual, expected)
 
 
+class EncryptionTests(TempfilesTestCase):
+    def test_roundtrip(self):
+        for encryptor in encryptors():
+            with self.subTest(encryptor=encryptor):
+                expected = b'some bytes'
+
+                path = self.given_tempfile(encryptor.extension)
+
+                with open(path, 'wb') as fobj:
+                    with encryptor.encrypt(fobj) as encrypted:
+                        encrypted.write(expected)
+
+                with open(path, 'rb') as fobj:
+                    self.assertNotEqual(fobj.read(), expected)
+
+                with open(path, 'rb') as fobj:
+                    with encryptor.deserialize(fobj) as decrypted:
+                        self.assertEqual(decrypted.read(), expected)
+
+
 if __name__ == '__main__':
     from unittest import main
     main()
```

However, this test fails since `fobj.read()` is the same as `expected` which I find surprising since the underlying file should have been encrypted, right? What am I doing wrong? Would it be possible for you to add a unit test that verifies the encryption behavior? Thanks in advance!

**Review by c-w — CHANGES_REQUESTED** [review]

File: benchmarks/compression.py [hunk 9]
```diff
@@ -23,6 +24,12 @@ def open_write(self, path: str) -> IO[bytes]:
     def open_read(self, path: str) -> IO[bytes]:
         raise NotImplementedError
 
+    def write_stream(self, fobj: IO[bytes]) -> IO[bytes]:
```
Comment (c-w) [inline benchmarks/compression.py]: This is a good refactor by the way to make the API more general. Great job! In order to enable this pull request to be more targeted and to ensure that all the tests also get updated, I've pull out this refactor into a separate PR: https://github.com/ascoderu/compression-benchmarks/pull/3

Comment (c-w) [reply benchmarks/compression.py]: I've merged the refactor with your branch and pushed the results.

File: .travis.yml (hunk 7 above)
Comment (c-w) [reply .travis.yml]: Done.

**Review by c-w — COMMENTED** [review]
Running the benchmarks, I get the following errors:

```
Error during write-phase in (none)+jsonl+aes: seek of closed file
Error during read-phase in (none)+jsonl+aes: 'NoneType' object has no attribute 'finalize'
Error during read-phase in (none)+bsonl+(none): objsize too large
Error during write-phase in (none)+bsonl+aes: seek of closed file
Error during read-phase in (none)+bsonl+aes: 'NoneType' object has no attribute 'finalize'
Error during write-phase in (none)+msgpack+aes: seek of closed file
Error during read-phase in (none)+msgpack+aes: 'NoneType' object has no attribute 'finalize'
Error during read-phase in (none)+avro+(none):
Error during write-phase in (none)+avro+aes: seek of closed file
Error during read-phase in (none)+avro+aes: 'NoneType' object has no attribute 'finalize'
Error during read-phase in gz+bsonl+(none): objsize too large
Error during read-phase in gz+bsonl+aes: objsize too large
Error during read-phase in gz+avro+(none):
Error during read-phase in gz+avro+aes:
Error during read-phase in 3.zs+bsonl+(none): objsize too large
Error during read-phase in 3.zs+bsonl+aes: objsize too large
Error during write-phase in 3.zs+avro+(none): 'zstd.ZstdCompressionWriter' object has no attribute 'seekable'
Error during read-phase in 3.zs+avro+(none): cannot read header - is it an avro file?
Error during write-phase in 3.zs+avro+aes: 'zstd.ZstdCompressionWriter' object has no attribute 'seekable'
Error during read-phase in 3.zs+avro+aes: 'NoneType' object has no attribute 'finalize'
Error during read-phase in 22.zs+bsonl+(none): objsize too large
Error during read-phase in 22.zs+bsonl+aes: objsize too large
Error during write-phase in 22.zs+avro+(none): 'zstd.ZstdCompressionWriter' object has no attribute 'seekable'
Error during read-phase in 22.zs+avro+(none): cannot read header - is it an avro file?
Error during write-phase in 22.zs+avro+aes: 'zstd.ZstdCompressionWriter' object has no attribute 'seekable'
Error during read-phase in 22.zs+avro+aes: 'NoneType' object has no attribute 'finalize'
Error during read-phase in tar.bz2+bsonl+(none): objsize too large
Error during read-phase in tar.bz2+bsonl+aes: objsize too large
Error during read-phase in tar.bz2+avro+(none):
Error during read-phase in tar.bz2+avro+aes:
Error during read-phase in tar.xz+bsonl+(none): objsize too large
Error during read-phase in tar.xz+bsonl+aes: objsize too large
Error during read-phase in tar.xz+avro+(none):
Error during read-phase in tar.xz+avro+aes:
```

I'm not too concerned about the `avro` or `bsonl` errors.

However, the error for `(none)+jsonl+aes` seems a bit odd. Could you take a look?

**Comment by adamsclafani** [comment]
Done. The msgpack errors should have been fixed as well.

**Review by c-w — APPROVED** [review]
Thanks again for the great work!
